core/teams/engine.go

```diff
@@ -0,0 +1,148 @@
+
+	// membersToMove tracks all the parties that switch teams. The switch only
+	// happens by the end of the epoch.
+	membersToMove map[types.PartyID]teamMove
+}
+
+func (e *Engine) CreateTeam(referrer types.PartyID, name string, teamURL string, avatarURL string, enableRewards bool) (types.TeamID, error) {
+	newTeamID := e.newUniqueTeamID()
+
+	if _, alreadyMember := e.allTeamMembers[referrer]; alreadyMember {
+		return "", fmt.Errorf("the party %q already belongs to a team", referrer)
+	}
+
+	e.teams[newTeamID] = &types.Team{
+		ID:            newTeamID,
+		Referrer:      referrer,
+		Name:          name,
+		TeamURL:       teamURL,
+		AvatarURL:     avatarURL,
+		EnableRewards: enableRewards,
+	}
+
+	e.allTeamMembers[referrer] = newTeamID
+
+	return newTeamID, nil
+}
+
+func (e *Engine) UpdateTeam(id types.TeamID, name string, teamURL string, avatarURL string, enableRewards bool) error {
+	_, exists := e.teams[id]
+	if !exists {
+		return errNoTeamMatchesID(id)
+	}
+
+	e.teams[id].Name = name
+	e.teams[id].TeamURL = teamURL
+	e.teams[id].AvatarURL = avatarURL
+	e.teams[id].EnableRewards = enableRewards
+
+	return nil
+}
+
+func (e *Engine) JoinTeam(id types.TeamID, party types.PartyID) error {
+	for _, team := range e.teams {
+		if team.Referrer == party {
+			return ErrReferrerCannotJoinAnotherTeam
+		}
+	}
+
+	teamToJoin, exists := e.teams[id]
+	if !exists {
+		return errNoTeamMatchesID(id)
+	}
+
+	teamJoined, alreadyMember := e.allTeamMembers[party]
+	if alreadyMember {
+		// This party is already member of a team, it will be moved at the end
+		// of epoch.
+		e.membersToMove[party] = teamMove{
+			fromTeam: teamJoined,
+			toTeam:   teamToJoin.ID,
+		}
+		return nil
+	}
+
+	// The party does not belong to a team, so he joins right away.
+	teamToJoin.AddReferee(party)
+
+	e.allTeamMembers[party] = teamToJoin.ID
+
+	return nil
+}
+
+func (e *Engine) OnEpoch(_ context.Context, ep types.Epoch) {
+	switch ep.Action {
+	case proto.EpochAction_EPOCH_ACTION_END:
+		e.moveMembers()
+	}
+}
+
+func (e *Engine) OnEpochRestore(_ context.Context, _ types.Epoch) {}
+
+func (e *Engine) moveMembers() {
+	for party, move := range e.membersToMove {
+		e.teams[move.fromTeam].RemoveReferee(party)
+		e.teams[move.toTeam].AddReferee(party)
+	}
+}
+
+func (e *Engine) newUniqueTeamID() types.TeamID {
+	for {
+		id := types.NewTeamID()
+		if _, exists := e.teams[id]; !exists {
+			return id
+		}
+	}
+}
+
+func (e *Engine) IsTeamMember(party types.PartyID) bool {
+	_, isMember := e.allTeamMembers[party]
+	return isMember
+}
+
+func NewEngine(epochEngine EpochEngine) *Engine {
+	engine := &Engine{
+		teams:          map[types.TeamID]*types.Team{},
+		allTeamMembers: map[types.PartyID]types.TeamID{},
+		membersToMove:  map[types.PartyID]teamMove{},
+	}
+
+	epochEngine.NotifyOnEpoch(engine.OnEpoch, engine.OnEpochRestore)
+
+	return engine
+}
+
+func errNoTeamMatchesID(id types.TeamID) error {
+	return fmt.Errorf("no team matches ID %q", id)
+}
```

Comment:
I usually create an error like this for this kind of things:
```suggestion
var ErrNoTeamMatchesID = func(id types.TeamID) error {
	return fmt.Errorf("no team matches ID %q", id)
}
```

So you can still export the error and use it in your test like this: `ErrNoTeamMatchesID("myteamid")`